src/interface/variable_pack.hpp

```diff
@@ -158,203 +255,181 @@ struct PackAndIndexMap {
+  Kokkos::deep_copy(cv_out, host_cv);
+  Kokkos::deep_copy(sparse_id_out, host_sp);
+  Kokkos::deep_copy(vector_component_out, host_vc);
 }
 
 template <typename T>
-void FillFluxViews(const vpack_types::VarList<T> &vars, PackIndexMap *vmap,
-                   const int ndim, ViewOfParArrays<T> &f1, ViewOfParArrays<T> &f2,
-                   ViewOfParArrays<T> &f3) {
+void FillFluxViews(const CellVariableVector<T> &vars, const int ndim,
+                   ViewOfParArrays<T> &f1_out, ViewOfParArrays<T> &f2_out,
+                   ViewOfParArrays<T> &f3_out, PackIndexMap *vmap_out) {
   using vpack_types::IndexPair;
 
-  auto host_f1 = Kokkos::create_mirror_view(Kokkos::HostSpace(), f1);
-  auto host_f2 = Kokkos::create_mirror_view(Kokkos::HostSpace(), f2);
-  auto host_f3 = Kokkos::create_mirror_view(Kokkos::HostSpace(), f3);
+  auto host_f1 = Kokkos::create_mirror_view(Kokkos::HostSpace(), f1_out);
+  auto host_f2 = Kokkos::create_mirror_view(Kokkos::HostSpace(), f2_out);
+  auto host_f3 = Kokkos::create_mirror_view(Kokkos::HostSpace(), f3_out);
 
   int vindex = 0;
-  int sparse_start = -22000; // Initialize to an obviously wrong number
-  std::string sparse_name;
   for (const auto &v : vars) {
-    if (vmap != nullptr) {
-      if (v->IsSet(Metadata::Sparse)) {
-        std::string sparse_trim = v->label();
-        sparse_trim.erase(sparse_trim.find_last_of("_"));
-        if (sparse_name == "") {
-          sparse_name = sparse_trim;
-          sparse_start = vindex;
-        }
-        if (sparse_name != sparse_trim) {
-          vmap->insert(std::pair<std::string, IndexPair>(
-              sparse_name, IndexPair(sparse_start, vindex - 1)));
-          sparse_name = sparse_trim;
-          sparse_start = vindex;
-        }
-      } else if (!(sparse_name == "")) {
-        vmap->insert(std::pair<std::string, IndexPair>(
-            sparse_name, IndexPair(sparse_start, vindex - 1)));
-        sparse_name = "";
-      }
-    }
     int vstart = vindex;
     for (int k = 0; k < v->GetDim(6); k++) {
       for (int j = 0; j < v->GetDim(5); j++) {
         for (int i = 0; i < v->GetDim(4); i++) {
-          host_f1(vindex) = v->flux[X1DIR].Get(k, j, i);
-          if (ndim >= 2) host_f2(vindex) = v->flux[X2DIR].Get(k, j, i);
-          if (ndim >= 3) host_f3(vindex) = v->flux[X3DIR].Get(k, j, i);
+          if (v->IsAllocated()) {
+            host_f1(vindex) = v->flux[X1DIR].Get(k, j, i);
+            if (ndim >= 2) host_f2(vindex) = v->flux[X2DIR].Get(k, j, i);
+            if (ndim >= 3) host_f3(vindex) = v->flux[X3DIR].Get(k, j, i);
+          } else {
+            host_f1(vindex) = ParArray3D<T>("unallocated_f1_" + v->label(), 0, 0, 0);
+            if (ndim >= 2) {
+              host_f2(vindex) = ParArray3D<T>("unallocated_f2_" + v->label(), 0, 0, 0);
+            }
+            if (ndim >= 3) {
+              host_f3(vindex) = ParArray3D<T>("unallocated_f3_" + v->label(), 0, 0, 0);
+            }
+          }
```

Comment:
I believe you are right. When testing this, I found another issue that broke mesh data packs, when some sparse fields are not allocated (which isn't happening yet in this PR, but I wanted to test it). That issue is fixed now and the calculate pi example (which doesn't do any communication) now works with sparsely allocated sparse fields (but that's not part of this PR).

Reply:
What was the issue?

Reply:
Basically, the dimension of the mesh data pack was based on the dimension of the first mesh block data in the pack. But if that happened to only contain unallocated fields, its dimension was 0 x 0 x 0, which then made the entire dimension of the mesh data pack 0 x 0 x 0 and so no computation would be done on the entire pack, even though other mesh block data objects in the pack had data allocated.

What I've changed is that the dimension of a `CellVariable` are now stored in the `CellVariable` directly, instead of querying the underlying `ParArrayND`, which may or may not be allocated and hence may or may not have the actual (logical) dimensions of the `CellVariable`. It is entirely possible, though, that there will be other corner cases that come up once we don't allocate all sparse fields anymore.